### PR TITLE
feat(selection): implement simple selection to bind data into DOM of @antvis/g for animation and interaction

### PR DESCRIPTION
**Pull request by pearmini**
Implement simple selection to bind data into DOM of @antvis/g for animation and interaction. (Test coverage: 96%)

**Features**：
- Statistic: Key
- Infer: MaybeKey

**Refactor**:
- Runtime: Render marks and components.
- Shape: Line, Smooth, Rect, HollowRect
- Geometry: Interval, Line

**Reference**

Geometry returns points and indexes for shape instead of shapes. This will make geometry more pure because it only computes points for each shape.

```ts
// Before
const shapes: DisplayObject[] = interval(index, scale, value, coordinate, theme);

// Now
const [I, P]: [number[], Vector2[][]] = interval(index, scale, value, coordinate);
```

Split value and style for shape and pass theme directly to shape. This will provide more flexibility to determinate the style of shape from value, style options and theme.

```ts
// Before
const shape = Rect()(P, style, coordinate);

// Now
const shape = Rect({ stroke: 'red' })(P, value, coordinate, theme);
```
Selection will bind data to DOM as following. Let's say we have structure as follow:

```html
<g>
  <rect className='cls' __data__='1'></rect>
  <rect className='cls' __data__='2'></rect>
  <rect className='cls' __data__='3'></rect>
</g>
```

If data have been changed, selection will diff new data and old elements, then produce *enter*, *update* and *exit* selection.

```js
const newData = [2, 3, 4];

const newSelection = selection
  .selectAll('.cls') // Select all elements with cls as className.
  .data(data) // Bind data. 

newSelection._enter // {_data: [4], _elements: []}
newSelection._update // {_data: null, _elements: [rect{2}, rect{3}]}
newSelection._exit // {_data: null, _elements: [rect{1}]}

const joinedSelection = newSelection.join(
  (enter) => enter.append('rect'),  // Append rect for 4
  (update) => update.style('fill', 'red'), // Update rect for  2 and 3
  (exit) => exit.remove(), // Remove rect for 1
);

joinedSelection // {_data: null, _elements: [rect{2}, rect{3}, rect{4}]};
```

The structure will result as follow.

```html
<g>
  <rect className='cls' __data__='2' style="fill: red"></rect>
  <rect className='cls' __data__='3' style="fill: red"></rect>
  <rect className='cls' __data__='4'></rect>
</g>
```

The more clear code is as follow.

```js
selection
  .selectAll('.cls')
  .data([2, 3, 4])
  .join(
    (enter) => enter.append('rect'), 
    (update) => update.style('fill', 'red'),
    (exit) => exit.remove(),
  );
```

### CODE REVIEW

File: src/utils/selection.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    return this;
+  }
+
+  node(): G2Element {
+    return this._elements[0];
+  }
+
+  nodes(): G2Element[] {
+    return this._elements;
+  }
```
Comment (hustcc) [inline src/utils/selection.ts:247]: 这就是你说的 selection？

Comment (pearmini) [reply src/utils/selection.ts:247]: 嗯嗯

File: __tests__/unit/shape/line/smooth.spec.ts [hunk 2]
```diff
@@ -32,7 +32,7 @@ describe('Smooth', () => {
     const shape = draw({
       shape: Smooth({ alpha: 0 }),
       container,
-      style: {
+      value: {
```
Comment (visiky) [inline __tests__/unit/shape/line/smooth.spec.ts:35]: 要不要改成 `values` 代表各通道中的映射数据

File: __tests__/unit/shape/interval/rect.spec.ts [hunk 3]
```diff
@@ -155,11 +152,10 @@ describe('Rect', () => {
   it('Rect() returns a function enabling drawing sector with radius in Polar', () => {
     const container = document.createElement('div');
     const shape = draw({
-      shape: Rect(),
+      shape: Rect({ radius: 20 }),
```
Comment (visiky) [inline __tests__/unit/shape/interval/rect.spec.ts:155]: [MARK] maybe consider using `style` as a channel later

**Review by visiky — APPROVED** [review]

File: __tests__/unit/shape/line/smooth.spec.ts (hunk 2 above)
Comment (pearmini) [reply __tests__/unit/shape/line/smooth.spec.ts:35]: 这里是一个对象？所以用单数？

**Pull request merged**